zilliqa/src/consensus.rs

```diff
@@ -1394,42 +1415,42 @@ impl Consensus {
         }
     }
 
-    fn check_and_commit(&mut self, proposal_hash: Hash) -> Result<()> {
-        let Some(proposal) = self.get_block(&proposal_hash)? else {
-            trace!("block not found: {proposal_hash}");
-            return Ok(());
+    fn check_and_commit(&mut self, proposal: &Block) -> Result<()> {
```

Comment:
The finalized logic here is rewritten to be actually correct. It is important for the block verification logic to work correctly (discarding/not voting on blocks that QC points before finalized). Note that the finalized is not HEAD-3 neccessarily if the views are not contigous.

Reply:
As Zoltan clarified the correct condition is actually more lax (the two-chain has to have consecutive views `n-k`, `n-k+1` but the latest block at view `n` doesn't have to be consecutive, i.e. `k >= 2`). I'm fine with fixing that in a subsequent PR if you prefer, up to you